### PR TITLE
feat(api): add fft

### PR DESCRIPTION
**Pull request by minghuaw**

### CODE REVIEW

File: src/stream.rs [hunk 1]
```diff
@@ -84,6 +84,7 @@ impl Stream {
         Stream { c_stream }
     }
 
+    // TODO: document how this is different from `Default::default()`
```
Comment (dcvz) [inline src/stream.rs:87]: We could name this `default_stream_for_device()`?

Comment (minghuaw) [reply src/stream.rs:87]: Sounds good. I will make that change in this PR

Comment (minghuaw) [reply src/stream.rs]: Renamed to `default_stream_on_device`

**Comment by dcvz** [comment]
@minghuaw I wonder if you can leverage your ported `fftn`, `ifftn` , `irfftn` and `rfftn` for  `fft2`, `ifft2`, `irfft2` and `rfft2`  - might make validations a tiny bit simpler eventhough you've already tried to generalize checks a lot 
internally it does something similar as well and it might allow you to have less inners

**Comment by dcvz** [comment]
It's all built on top of the "n" variants with default axis/axes: https://github.com/ml-explore/mlx/blob/d0dbfe0b97b95bc29009e7534f7ff734fb679801/mlx/fft.h#L120

File: src/fft/fftn.rs [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+    try_resolve_sizes_and_axes,
+};
+
+/// One dimensional discrete Fourier Transform.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `n`: Size of the transformed axis. The corresponding axis in the input is truncated or padded
+///   with zeros to match `n`. The default value is `a.shape[axis]`.
```
Comment (dcvz) [inline src/fft/fftn.rs:16]:
```suggestion
///   with zeros to match `s`. The default value is `a.shape[axis]`.
```

File: src/fft/fftn.rs [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+        Array::from_ptr(c_array)
+    }
+}
+
+/// One dimensional discrete Fourier Transform.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `n`: Size of the transformed axis. The corresponding axis in the input is truncated or padded
```
Comment (dcvz) [inline src/fft/fftn.rs:37]: for some you call this `n` and for some `s`

File: src/fft/fftn.rs [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+    try_fft_device(a, n, axis, stream).unwrap()
+}
+
+/// Two dimensional discrete Fourier Transform.
+///
+/// # Param
+///
+/// - `a`: The input array.
+/// - `s`: Size of the transformed axes. The corresponding axes in the input are truncated or padded
+///  with zeros to match `n`. The default value is the sizes of `a` along `axes`.
```
Comment (dcvz) [inline src/fft/fftn.rs:81]:
```suggestion
///  with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+    fftn_device_unchecked(a, s, axes, stream)
+}
+
+/// Two dimensional discrete Fourier Transform.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `s`: Size of the transformed axes. The corresponding axes in the input are truncated or padded
+/// with zeros to match `n`. The default value is the sizes of `a` along `axes`.
```
Comment (dcvz) [inline src/fft/fftn.rs:100]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+    try_fftn_device(a, s, axes, stream)
+}
+
+/// Two dimensional discrete Fourier Transform.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `s`: Size of the transformed axes. The corresponding axes in the input are truncated or padded
+/// with zeros to match `n`. The default value is the sizes of `a` along `axes`.
```
Comment (dcvz) [inline src/fft/fftn.rs:119]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+    try_ifft_device(a, n, axis, stream).unwrap()
+}
+
+/// Two dimensional inverse discrete Fourier Transform.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `s`: Size of the transformed axes. The corresponding axes in the input are truncated or padded
+/// with zeros to match `n`. The default value is the sizes of `a` along `axes`.
```
Comment (dcvz) [inline src/fft/fftn.rs:288]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+    ifftn_device_unchecked(a, s, axes, stream)
+}
+
+/// Two dimensional inverse discrete Fourier Transform.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `s`: Size of the transformed axes. The corresponding axes in the input are truncated or padded
+/// with zeros to match `n`. The default value is the sizes of `a` along `axes`.
```
Comment (dcvz) [inline src/fft/fftn.rs:307]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/fftn.rs [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+    try_ifftn_device(a, s, axes, stream)
+}
+
+/// Two dimensional inverse discrete Fourier Transform.
+///
+/// # Params
+///
+/// - `a`: The input array.
+/// - `s`: Size of the transformed axes. The corresponding axes in the input are truncated or padded
+/// with zeros to match `n`. The default value is the sizes of `a` along `axes`.
```
Comment (dcvz) [inline src/fft/fftn.rs:326]:
```suggestion
/// with zeros to match `s`. The default value is the sizes of `a` along `axes`.
```

File: src/fft/mod.rs [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+use crate::{
+    error::FftError,
+    utils::{all_unique, resolve_index, resolve_index_unchecked},
+    Array,
+};
+
+pub use self::{fftn::*, rfftn::*};
+
+#[inline]
+fn resolve_size_and_axis_unchecked(
```
Comment (dcvz) [inline src/fft/mod.rs:184]: minor: I'd love to move these to an `fft/utils.rs` just to make this module easy to digest for those looking.

File: src/fft/fftn.rs [hunk 11]
```diff
@@ -0,0 +1,534 @@
...
+    a: &'a Array,
+    s: impl Into<Option<&'a [i32]>>,
+    axes: impl Into<Option<&'a [i32]>>,
+    stream: StreamOrDevice,
+) -> Array {
+    try_ifftn_device(a, s, axes, stream).unwrap()
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (dcvz) [inline src/fft/fftn.rs:423]: I like to add tests for known failures -- but if you think its enough to be covered by the tests in the utils you're using to validate, then I leave it up to you.

**Review by dcvz — APPROVED** [review]
LGTM 👍 one suggestion for the validations and some inconsistencies with docs (i didn't list them all) but otherwise this LGTM. Approving so you can merge when you've resolved comments.

File: src/error.rs [hunk 12]
```diff
@@ -51,3 +51,21 @@ pub enum AsSliceError {
...
+
+#[derive(Error, Debug, PartialEq)]
+pub enum FftError {
+    #[error("fftn requires at least one dimension")]
+    ScalarArray,
+
+    #[error("Invalid axis received for array with {ndim} dimensions")]
+    InvalidAxis { ndim: usize },
+
+    #[error("Shape and axes/axis have different sizes")]
```
[the remark this answers is not in the document]
Comment (minghuaw) [reply src/error.rs:63]: Didn't do this because the c++ exception's not doing it

File: src/fft/fftn.rs (hunk 2 above)
Comment (minghuaw) [reply src/fft/fftn.rs:16]: This follows the python api. 1D fft uses `n`, n-D uses `s`

File: src/fft/fftn.rs (hunk 3 above)
Comment (minghuaw) [reply src/fft/fftn.rs:37]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs (hunk 4 above)
Comment (minghuaw) [reply src/fft/fftn.rs:81]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs (hunk 5 above)
Comment (minghuaw) [reply src/fft/fftn.rs:100]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs (hunk 6 above)
Comment (minghuaw) [reply src/fft/fftn.rs:119]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs (hunk 7 above)
Comment (minghuaw) [reply src/fft/fftn.rs:288]: This follows the python api. 1D fft uses n, n-D uses s

File: src/fft/fftn.rs (hunk 5 above)
Comment (dcvz) [reply src/fft/fftn.rs:100]: In this case though, you used `s` as a name but then used `n` here. so there's a mismatch

Comment (minghuaw) [reply src/fft/fftn.rs:100]: Yeah, didn't realize that

File: src/fft/fftn.rs (hunk 11 above)
Comment (minghuaw) [reply src/fft/fftn.rs:423]: I think this would suffice

**Pull request merged**